File: aoj/src/GRL_3_A/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
)

var sc = bufio.NewScanner(os.Stdin)
var adj [][]int
var prenum, parent, lowest, out []int
var visited []bool
var timer = 1

func dfs(current, prev int) {

	prenum[current] = timer
	lowest[current] = timer
	timer++
	visited[current] = true

	for _, v := range adj[current] {
		if !visited[v] {
			parent[v] = current

			dfs(v, current)
			if lowest[current] > lowest[v] {
				lowest[current] = lowest[v]
			}
		} else if v != prev {
			if lowest[current] > prenum[v] {
				lowest[current] = prenum[v]
			}
		}
	}
}

// GRL_3_A: 関節点
func main() {
	sc.Buffer([]byte{}, math.MaxInt64)
	sc.Split(bufio.ScanWords)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	v := nextInt()
	e := nextInt()

	adj = make([][]int, v, v)
	prenum = make([]int, v, v)
	parent = make([]int, v, v)
	lowest = make([]int, v, v)
	visited = make([]bool, v, v)
	out = []int{}

	for i := 0; i < v; i++ {
		adj[i] = make([]int, 0, 0)
	}

	for i := 0; i < e; i++ {
		src := nextInt()
		dst := nextInt()
		adj[src] = append(adj[src], dst)
		adj[dst] = append(adj[dst], src)
	}

	dfs(0, 0)

	np := 0
	for i := 1; i < v; i++ {
		p := parent[i]
		if p == 0 {
			np++
		} else {
			if prenum[p] <= lowest[i] {
				out = append(out, p)
			}
		}
	}

	if np > 1 {
		out = append(out, 0)
	}

	sort.Sort(sort.IntSlice(out))

	prev := -1
	for _, v := range out {
		if v == prev {
			continue
		}
		fmt.Fprintln(w, v)
		prev = v
	}
}

func next() string {
	sc.Scan()
	return sc.Text()
}

func nextInt() int {
	s := next()
	i, _ := strconv.Atoi(s)
	return i
}

File: aoj/src/GRL_3_A/main.py
import sys

class Resolver:
    def __init__(self, adj):
        self.adj = adj
        self.t = 1
        self.visited = [False] * len(adj)
        self.parent = [0] * len(adj)
        self.pre = [0] * len(adj)
        self.lowest = [0] * len(adj)

    def resolve(self):
        self.dfs(0, 0)
        ret = []
        np = False
        for i in range(1, len(self.adj)):
            p = self.parent[i]
            if p == 0:
                np = True
            else:
                if self.pre[p] <= self.lowest[i]:
                    ret.append(p)
        if np:
            ret.append(0)
        ret = list(set(ret))
        ret.sort()
        return ret


    def dfs(self, i, previous_i):
        self.visited[i] = True
        self.pre[i] = self.t
        self.lowest[i] = self.t
        self.t += 1

        for v in self.adj[i]:
            if not self.visited[v]:
                self.parent[v] = i
                self.dfs(v, i)
                if self.lowest[i] > self.lowest[v]:
                    self.lowest[i] = self.lowest[v]
            elif v != previous_i:
                if self.lowest[i] > self.pre[v]:
                    self.lowest[i] = self.pre[v]

def main():
    readline = sys.stdin.readline
    v, e = map(int, readline().split())
    adj = [[] for _ in range(v)]
    for _ in range(e):
        s, d = map(int, readline().split())
        adj[s].append(d)
        adj[d].append(s)
    result = Resolver(adj).resolve()
    print(*result, sep="\n")

if __name__ == "__main__":
    main()
